### DISCUSSION TITLE
How can I debug compute pass scheduling?

### DISCUSSION THREAD

**Discussion by Patryk27** [gfx-rs/wgpu#5074] [Q&A]
Hi,

I've got a couple of compute shaders that work on two different buffers and later some another shader merges those two buffers together - I've got a strong suspicion that most of those shaders (being mixed ALU/mem-bound) should be scheduled in parallel, but it looks like Mac only "parallelizes" the first ones:

<img width="889" alt="image" src="https://github.com/gfx-rs/wgpu/assets/3395477/04e753b1-b324-4b63-8f85-1f1c43f159b6">

For completion, that's how it looks like in Xcode's dependencies tab:

<img width="269" alt="image" src="https://github.com/gfx-rs/wgpu/assets/3395477/bca809e9-375f-4aba-86f4-5ae5b923c785">

Inspecting those shaders tells me that they spend a lot of time waiting for memory (and don't actually use that many registers), making them a perfect candidate for parallelization:

<img width="1125" alt="image" src="https://github.com/gfx-rs/wgpu/assets/3395477/6e5e3e6a-66ce-42a9-b075-b5c081b7f2f4">

... but it's not happening - is there any way for me to debug why scheduler would do this or the other thing?

(maybe there's some low-hanging fruit to check, some accidentally-created synchronization edge that I could get rid of etc.)

Thanks!

Edit: of course I'm aware that passes being executed in parallel doesn't imply any performance gain (and that it's some magical code in the driver which ultimately decides what's happening) - I'd just like to make sure there's no low-hanging optimization-fruit there.

**Comment by cwfitzgerald** [comment]
Thankfully, this is very simple! They are defined to run in serial in WebGPU. Each dispatch is its own usage scope, so will run as-if serial. We communicate this to metal by making a "serial" compute pass, causing this behavior.

See #2659 for discussion of parallel compute passes - something we want just haven't been implemented

**Reply by Patryk27** [reply]
Yes, but that's within a single command buffers, no? Compute passes across command buffers can still be parallelized by the driver (as seen in the image above), right? 🤔

**Reply by Patryk27** [reply]
Or, to put it differently, if currently the implementation is force-serial, then why I see two passes getting executed concurrently above? 😅

**Reply by Patryk27** [reply]
For comparison, here's how it looks like on the waterfall view in Xcode 👀
(the tiny slices being, it seems, the allocated time slots for given shader)

<img width="731" alt="image" src="https://github.com/gfx-rs/wgpu/assets/3395477/8a381670-cdaa-4b28-971a-d182613831a5">

**Reply by Patryk27** [reply]
cc @cwfitzgerald 👀

**Reply by BGR360** [reply]
> Compute passes across command buffers can still be parallelized by the driver (as seen in the image above), right?

@Patryk27 no. I recently asked a similar question and found out that [command buffers execute serially](https://github.com/gfx-rs/wgpu/discussions/6268#discussioncomment-10644147).
